// @vercel/sandbox/dist/api-client/validators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnapshotResponse = exports.CreateSnapshotResponse = exports.UpdateNetworkPolicyResponse = exports.ExtendTimeoutResponse = exports.SnapshotsResponse = exports.SandboxesResponse = exports.LogLine = exports.LogError = exports.LogLineStderr = exports.LogLineStdout = exports.EmptyResponse = exports.CommandFinishedResponse = exports.CommandResponse = exports.SandboxAndRoutesResponse = exports.SandboxResponse = exports.Command = exports.Pagination = exports.Snapshot = exports.SandboxRoute = exports.Sandbox = exports.NetworkPolicyValidator = exports.InjectionRuleValidator = void 0;
const zod_1 = require("zod");
exports.InjectionRuleValidator = zod_1.z.object({
    domain: zod_1.z.string(),
    // headers are only sent in requests
    headers: zod_1.z.record(zod_1.z.string()).optional(),
    // headerNames are returned in responses
    headerNames: zod_1.z.array(zod_1.z.string()).optional(),
});
exports.NetworkPolicyValidator = zod_1.z.union([
    zod_1.z.object({ mode: zod_1.z.literal("allow-all") }).passthrough(),
    zod_1.z.object({ mode: zod_1.z.literal("deny-all") }).passthrough(),
    zod_1.z
        .object({
        mode: zod_1.z.literal("custom"),
        allowedDomains: zod_1.z.array(zod_1.z.string()).optional(),
        allowedCIDRs: zod_1.z.array(zod_1.z.string()).optional(),
        deniedCIDRs: zod_1.z.array(zod_1.z.string()).optional(),
        injectionRules: zod_1.z.array(exports.InjectionRuleValidator).optional(),
    })
        .passthrough(),
]);
exports.Sandbox = zod_1.z.object({
    id: zod_1.z.string(),
    memory: zod_1.z.number(),
    vcpus: zod_1.z.number(),
    region: zod_1.z.string(),
    runtime: zod_1.z.string(),
    timeout: zod_1.z.number(),
    status: zod_1.z.enum([
        "pending",
        "running",
        "stopping",
        "stopped",
        "failed",
        "aborted",
        "snapshotting",
    ]),
    requestedAt: zod_1.z.number(),
    startedAt: zod_1.z.number().optional(),
    requestedStopAt: zod_1.z.number().optional(),
    stoppedAt: zod_1.z.number().optional(),
    abortedAt: zod_1.z.number().optional(),
    duration: zod_1.z.number().optional(),
    sourceSnapshotId: zod_1.z.string().optional(),
    snapshottedAt: zod_1.z.number().optional(),
    createdAt: zod_1.z.number(),
    cwd: zod_1.z.string(),
    updatedAt: zod_1.z.number(),
    interactivePort: zod_1.z.number().optional(),
    networkPolicy: exports.NetworkPolicyValidator.optional(),
    activeCpuDurationMs: zod_1.z.number().optional(),
    networkTransfer: zod_1.z.object({
        ingress: zod_1.z.number(),
        egress: zod_1.z.number(),
    }).optional(),
});
exports.SandboxRoute = zod_1.z.object({
    url: zod_1.z.string(),
    subdomain: zod_1.z.string(),
    port: zod_1.z.number(),
});
exports.Snapshot = zod_1.z.object({
    id: zod_1.z.string(),
    sourceSandboxId: zod_1.z.string(),
    region: zod_1.z.string(),
    status: zod_1.z.enum(["created", "deleted", "failed"]),
    sizeBytes: zod_1.z.number(),
    expiresAt: zod_1.z.number().optional(),
    createdAt: zod_1.z.number(),
    updatedAt: zod_1.z.number(),
});
exports.Pagination = zod_1.z.object({
    /**
     * Amount of items in the current page.
     * @example 20
     */
    count: zod_1.z.number(),
    /**
     * Timestamp that must be used to request the next page.
     * @example 1540095775951
     */
    next: zod_1.z.number().nullable(),
    /**
     * Timestamp that must be used to request the previous page.
     * @example 1540095775951
     */
    prev: zod_1.z.number().nullable(),
});
exports.Command = zod_1.z.object({
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    args: zod_1.z.array(zod_1.z.string()),
    cwd: zod_1.z.string(),
    sandboxId: zod_1.z.string(),
    exitCode: zod_1.z.number().nullable(),
    startedAt: zod_1.z.number(),
});
const CommandFinished = exports.Command.extend({
    exitCode: zod_1.z.number(),
});
exports.SandboxResponse = zod_1.z.object({
    sandbox: exports.Sandbox,
});
exports.SandboxAndRoutesResponse = exports.SandboxResponse.extend({
    routes: zod_1.z.array(exports.SandboxRoute),
});
exports.CommandResponse = zod_1.z.object({
    command: exports.Command,
});
exports.CommandFinishedResponse = zod_1.z.object({
    command: CommandFinished,
});
exports.EmptyResponse = zod_1.z.object({});
const LogLineBase = zod_1.z.object({ data: zod_1.z.string() });
exports.LogLineStdout = LogLineBase.extend({
    stream: zod_1.z.literal("stdout"),
});
exports.LogLineStderr = LogLineBase.extend({
    stream: zod_1.z.literal("stderr"),
});
exports.LogError = zod_1.z.object({
    stream: zod_1.z.literal("error"),
    data: zod_1.z.object({
        code: zod_1.z.string(),
        message: zod_1.z.string(),
    }),
});
exports.LogLine = zod_1.z.discriminatedUnion("stream", [
    exports.LogLineStdout,
    exports.LogLineStderr,
    exports.LogError,
]);
exports.SandboxesResponse = zod_1.z.object({
    sandboxes: zod_1.z.array(exports.Sandbox),
    pagination: exports.Pagination,
});
exports.SnapshotsResponse = zod_1.z.object({
    snapshots: zod_1.z.array(exports.Snapshot),
    pagination: exports.Pagination,
});
exports.ExtendTimeoutResponse = zod_1.z.object({
    sandbox: exports.Sandbox,
});
exports.UpdateNetworkPolicyResponse = zod_1.z.object({
    sandbox: exports.Sandbox,
});
exports.CreateSnapshotResponse = zod_1.z.object({
    snapshot: exports.Snapshot,
    sandbox: exports.Sandbox,
});
exports.SnapshotResponse = zod_1.z.object({
    snapshot: exports.Snapshot,
});
