// @vercel/prepare-flags-definitions/dist/index.js
// src/index.ts
import { createHash } from "crypto";
import { mkdir, writeFile } from "fs/promises";
import { join } from "path";

// package.json
var version = "0.2.1";

// src/index.ts
var FLAGS_HOST = "https://flags.vercel.com";
var FLAGS_DEFINITIONS_VERSION = "1.0.1";
function obfuscate(sdkKey, prefixLength = 18) {
  if (prefixLength >= sdkKey.length) return sdkKey;
  return sdkKey.slice(0, prefixLength) + "*".repeat(sdkKey.length - prefixLength);
}
function hashSdkKey(sdkKey) {
  return createHash("sha256").update(sdkKey).digest("hex");
}
var SDK_KEY_REGEX = /^vf_(?:server|client)_/;
function generateDefinitionsModule(sdkKeys, values) {
  const stringified = sdkKeys.map((_, i) => JSON.stringify(values[i]));
  const uniqueStrings = [];
  const stringToIndex = /* @__PURE__ */ new Map();
  for (const s of stringified) {
    if (!stringToIndex.has(s)) {
      stringToIndex.set(s, uniqueStrings.length);
      uniqueStrings.push(s);
    }
  }
  const keyToIndex = sdkKeys.map(
    (_, i) => stringToIndex.get(stringified[i]) ?? 0
  );
  const hashedKeys = sdkKeys.map(hashSdkKey);
  const lines = [
    "const memo = (fn) => { let cached; return () => (cached ??= fn()); };",
    ""
  ];
  for (let i = 0; i < uniqueStrings.length; i++) {
    lines.push(
      `const _d${i} = memo(() => JSON.parse(${JSON.stringify(uniqueStrings[i])}));`
    );
  }
  lines.push("");
  lines.push("const map = {");
  for (let i = 0; i < sdkKeys.length; i++) {
    lines.push(`  ${JSON.stringify(hashedKeys[i])}: _d${keyToIndex[i]},`);
  }
  lines.push("};");
  lines.push("");
  lines.push("export function get(hashedSdkKey) {");
  lines.push("  return map[hashedSdkKey]?.() ?? null;");
  lines.push("}");
  lines.push("");
  lines.push(
    `export const version = ${JSON.stringify(FLAGS_DEFINITIONS_VERSION)};`
  );
  return lines.join("\n");
}
async function prepareFlagsDefinitions(options) {
  const {
    cwd,
    env,
    userAgentSuffix,
    fetch: fetchFn = globalThis.fetch,
    output
  } = options;
  output == null ? void 0 : output.debug("vercel-flags: checking env vars for SDK Keys");
  const sdkKeys = Array.from(
    Object.values(env).reduce((acc, value) => {
      if (typeof value === "string") {
        if (SDK_KEY_REGEX.test(value)) {
          acc.add(value);
        } else if (value.startsWith("flags:")) {
          const params = new URLSearchParams(value.slice("flags:".length));
          const sdkKey = params.get("sdkKey");
          if (sdkKey && SDK_KEY_REGEX.test(sdkKey)) {
            acc.add(sdkKey);
          }
        }
      }
      return acc;
    }, /* @__PURE__ */ new Set())
  );
  output == null ? void 0 : output.debug(`vercel-flags: found ${sdkKeys.length} SDK keys`);
  if (sdkKeys.length === 0) {
    return { created: false, reason: "no-sdk-keys" };
  }
  const fetchPromise = Promise.all(
    sdkKeys.map(async (sdkKey) => {
      const headers = {
        authorization: `Bearer ${sdkKey}`,
        "user-agent": [
          `@vercel/prepare-flags-definitions/${version}`,
          userAgentSuffix
        ].filter(Boolean).join(" ")
      };
      if (env.VERCEL_PROJECT_ID) {
        headers["x-vercel-project-id"] = env.VERCEL_PROJECT_ID;
      }
      if (env.VERCEL_ENV) {
        headers["x-vercel-env"] = env.VERCEL_ENV;
      }
      if (env.VERCEL_DEPLOYMENT_ID) {
        headers["x-vercel-deployment-id"] = env.VERCEL_DEPLOYMENT_ID;
      }
      if (env.VERCEL_REGION) {
        headers["x-vercel-region"] = env.VERCEL_REGION;
      }
      const res = await fetchFn(`${FLAGS_HOST}/v1/datafile`, { headers });
      if (!res.ok) {
        throw new Error(
          `Failed to fetch flag definitions for ${obfuscate(sdkKey)}: ${res.status} ${res.statusText}`
        );
      }
      return res.json();
    })
  );
  const values = output ? await output.time("vercel-flags: load datafiles", fetchPromise) : await fetchPromise;
  const definitionsJs = generateDefinitionsModule(sdkKeys, values);
  const storageDir = join(cwd, "node_modules", "@vercel", "flags-definitions");
  const indexPath = join(storageDir, "index.js");
  const dtsPath = join(storageDir, "index.d.ts");
  const packageJsonPath = join(storageDir, "package.json");
  const dts = [
    "export function get(hashedSdkKey: string): Record<string, unknown> | null;",
    "export const version: string;",
    ""
  ].join("\n");
  const packageJson = {
    name: "@vercel/flags-definitions",
    version: FLAGS_DEFINITIONS_VERSION,
    type: "module",
    main: "./index.js",
    types: "./index.d.ts",
    exports: {
      ".": {
        types: "./index.d.ts",
        import: "./index.js"
      }
    }
  };
  await mkdir(storageDir, { recursive: true });
  await Promise.all([
    writeFile(indexPath, definitionsJs),
    writeFile(dtsPath, dts),
    writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2))
  ]);
  output == null ? void 0 : output.debug("vercel-flags: created module");
  output == null ? void 0 : output.debug(`  \u2192 ${indexPath}`);
  output == null ? void 0 : output.debug(`  \u2192 ${dtsPath}`);
  output == null ? void 0 : output.debug(`  \u2192 ${packageJsonPath}`);
  output == null ? void 0 : output.debug(
    `  \u2192 included definitions for keys "${sdkKeys.map((key) => obfuscate(key)).join(", ")}"`
  );
  return { created: true, sdkKeysCount: sdkKeys.length };
}
export {
  generateDefinitionsModule,
  hashSdkKey,
  prepareFlagsDefinitions
};
