// @vercel/sandbox/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.APIError = exports.StreamError = exports.CommandFinished = exports.Command = exports.Snapshot = exports.Sandbox = void 0;
var sandbox_1 = require("./sandbox");
Object.defineProperty(exports, "Sandbox", { enumerable: true, get: function () { return sandbox_1.Sandbox; } });
var snapshot_1 = require("./snapshot");
Object.defineProperty(exports, "Snapshot", { enumerable: true, get: function () { return snapshot_1.Snapshot; } });
var command_1 = require("./command");
Object.defineProperty(exports, "Command", { enumerable: true, get: function () { return command_1.Command; } });
Object.defineProperty(exports, "CommandFinished", { enumerable: true, get: function () { return command_1.CommandFinished; } });
var api_error_1 = require("./api-client/api-error");
Object.defineProperty(exports, "StreamError", { enumerable: true, get: function () { return api_error_1.StreamError; } });
var api_error_2 = require("./api-client/api-error");
Object.defineProperty(exports, "APIError", { enumerable: true, get: function () { return api_error_2.APIError; } });
