// @vercel/backends/dist/index.mjs
import { builtinModules, createRequire } from "node:module";
import { createWriteStream, existsSync, mkdirSync, mkdtempSync, readFileSync, rmSync, unlinkSync, writeFileSync } from "node:fs";
import { basename, dirname, extname, isAbsolute, join, posix, relative, resolve, sep } from "node:path";
import { delimiter, dirname as dirname$1, join as join$1 } from "path";
import { FileBlob, FileFsRef, NodejsLambda, Span, createDiagnostics, debug, defaultCachePathGlob, download, execCommand, generateProjectManifest, getEnvForPackageManager, getInternalServiceCronPath, getLambdaOptionsFromFunction, getNodeBinPaths, getNodeVersion, getReportedServiceType, glob, isBackendFramework, isBunVersion, isExperimentalBackendsWithoutIntrospectionEnabled, isScheduleTriggeredService, runNpmInstall, runPackageJsonScript, scanParentDirs } from "@vercel/build-utils";
import { lstat, readFile, rm, stat } from "node:fs/promises";
import { build as build$2 } from "rolldown";
import { exports } from "resolve.exports";
import { isNativeError } from "node:util/types";
import { nodeFileTrace as nodeFileTrace$1, resolve as resolve$1 } from "@vercel/nft";
import { transform } from "oxc-transform";
import execa from "execa";
import { readFile as readFile$1, writeFile } from "fs/promises";
import { fileURLToPath } from "node:url";
import { spawn } from "node:child_process";
import { tmpdir } from "node:os";
import { z } from "zod";

//#region src/utils.ts
async function downloadInstallAndBundle(args) {
	const { entrypoint, files, workPath, meta, config, repoRootPath } = args;
	await download(files, workPath, meta);
	const entrypointFsDirname = join$1(workPath, dirname$1(entrypoint));
	const { cliType, lockfilePath, lockfileVersion, packageJsonPackageManager, turboSupportsCorepackHome } = await scanParentDirs(entrypointFsDirname, true, repoRootPath);
	const spawnEnv = getEnvForPackageManager({
		cliType,
		lockfileVersion,
		packageJsonPackageManager,
		env: process.env,
		turboSupportsCorepackHome,
		projectCreatedAt: config.projectSettings?.createdAt
	});
	const installCommand = config.projectSettings?.installCommand;
	if (typeof installCommand === "string") if (installCommand.trim()) {
		console.log(`Running "install" command: \`${installCommand}\`...`);
		await execCommand(installCommand, {
			env: spawnEnv,
			cwd: entrypointFsDirname
		});
	} else console.log(`Skipping "install" command...`);
	else await runNpmInstall(entrypointFsDirname, [], { env: spawnEnv }, meta, config.projectSettings?.createdAt);
	return {
		entrypointFsDirname,
		spawnEnv,
		cliType,
		lockfilePath,
		lockfileVersion
	};
}
async function maybeExecBuildCommand(args, { spawnEnv }) {
	const projectBuildCommand = args.config.projectSettings?.buildCommand;
	if (projectBuildCommand) {
		const nodeBinPath = getNodeBinPaths({
			base: args.repoRootPath || args.workPath,
			start: args.workPath
		}).join(delimiter);
		return execCommand(projectBuildCommand, {
			env: {
				...spawnEnv,
				PATH: `${nodeBinPath}${delimiter}${spawnEnv?.PATH || process.env.PATH}`
			},
			cwd: args.workPath
		});
	}
	return runPackageJsonScript(args.workPath, ["build"], { env: spawnEnv }, args.config.projectSettings?.createdAt);
}

//#endregion
//#region src/diagnostics.ts
const diagnostics = createDiagnostics("node");

//#endregion
//#region src/cervel/plugin.ts
const CJS_SHIM_PREFIX$1 = "\0cjs-shim:";
const plugin = (args) => {
	const packageJsonCache = /* @__PURE__ */ new Map();
	const shimMeta = /* @__PURE__ */ new Map();
	const { tracedPaths } = args.context;
	const isBareImport = (id) => {
		return !id.startsWith(".") && !id.startsWith("/") && !/^[a-z][a-z0-9+.-]*:/i.test(id);
	};
	/**
	* Read and cache package.json contents
	*/
	const getPackageJson = async (pkgPath) => {
		if (packageJsonCache.has(pkgPath)) return packageJsonCache.get(pkgPath);
		try {
			const contents = await readFile(pkgPath, "utf-8");
			const parsed = JSON.parse(contents);
			packageJsonCache.set(pkgPath, parsed);
			return parsed;
		} catch {
			packageJsonCache.set(pkgPath, null);
			return null;
		}
	};
	/**
	* Determine if a resolved module is CommonJS based on package.json exports
	*/
	const isCommonJS = async (bareImport, resolvedPath, resolvedInfo) => {
		const ext = extname(resolvedPath);
		if (ext === ".cjs") return true;
		if (ext === ".mjs") return false;
		if (ext === ".js" || ext === ".ts") {
			const pkgJsonPath = resolvedInfo.packageJsonPath;
			if (!pkgJsonPath) return true;
			const pkgJson = await getPackageJson(pkgJsonPath);
			if (!pkgJson) return true;
			const pkgDir = dirname(pkgJsonPath);
			const relativePath = resolvedPath.startsWith(pkgDir) ? resolvedPath.slice(pkgDir.length + 1).replace(/\\/g, "/") : null;
			if (!relativePath) return pkgJson.type !== "module";
			const pkgName = pkgJson.name || "";
			const subpath = bareImport.startsWith(pkgName) ? `.${bareImport.slice(pkgName.length)}` || "." : ".";
			try {
				if (exports(pkgJson, subpath, {
					require: false,
					conditions: ["node", "import"]
				})?.some((p) => p === relativePath || p === `./${relativePath}`)) return false;
				if (exports(pkgJson, subpath, {
					require: true,
					conditions: ["node", "require"]
				})?.some((p) => p === relativePath || p === `./${relativePath}`)) return true;
			} catch (err) {
				console.warn("Export resolution failed::", err);
			}
			if (pkgJson.module) return false;
			return pkgJson.type !== "module";
		}
		return true;
	};
	const isLocalImport = (id) => {
		if (id.startsWith("node:")) return false;
		if (id.includes("node_modules")) return false;
		return true;
	};
	return {
		name: "cervel",
		resolveId: {
			order: "pre",
			async handler(id, importer, rOpts) {
				if (id.startsWith(CJS_SHIM_PREFIX$1)) return {
					id,
					external: false
				};
				const resolved = await this.resolve(id, importer, rOpts);
				if (builtinModules.includes(id)) return {
					id: `node:${id}`,
					external: true
				};
				if (resolved?.id && isLocalImport(resolved.id)) tracedPaths.add(resolved.id);
				if (importer?.startsWith(CJS_SHIM_PREFIX$1) && isBareImport(id)) return {
					id,
					external: true
				};
				if (importer && isBareImport(id) && resolved?.id?.includes("node_modules")) {
					if (args.shimBareImports) {
						if (await isCommonJS(id, resolved.id, resolved)) {
							const importerPkgJsonPath = (await this.resolve(importer))?.packageJsonPath;
							if (importerPkgJsonPath) {
								const importerPkgDir = relative(args.repoRootPath, dirname(importerPkgJsonPath));
								const shimId$1 = `${CJS_SHIM_PREFIX$1}${importerPkgDir.replace(/\//g, "_")}_${id.replace(/\//g, "_")}`;
								shimMeta.set(shimId$1, {
									pkgDir: importerPkgDir,
									pkgName: id
								});
								return {
									id: shimId$1,
									external: false
								};
							}
							const shimId = `${CJS_SHIM_PREFIX$1}${id.replace(/\//g, "_")}`;
							shimMeta.set(shimId, {
								pkgDir: "",
								pkgName: id
							});
							return {
								id: shimId,
								external: false
							};
						}
					}
					return {
						external: true,
						id
					};
				}
				if (importer && isBareImport(id)) return resolved;
				return {
					external: true,
					...resolved,
					id: resolved?.id || id
				};
			}
		},
		load: { async handler(id) {
			if (id.startsWith(CJS_SHIM_PREFIX$1)) {
				const meta = shimMeta.get(id);
				if (!meta) return { code: `module.exports = require('${id.slice(10)}');` };
				const { pkgDir, pkgName } = meta;
				if (pkgDir) return { code: `
import { createRequire } from 'node:module';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

const requireFromContext = createRequire(join(dirname(fileURLToPath(import.meta.url)), '${join("..", pkgDir, "package.json")}'));
module.exports = requireFromContext('${pkgName}');
`.trim() };
				return { code: `module.exports = require('${pkgName}');` };
			}
			return null;
		} }
	};
};

//#endregion
//#region src/cervel/node-file-trace.ts
const nodeFileTrace = async (args) => {
	const { span } = args;
	const files = {};
	const { tracedPaths } = args;
	const compiledSourceFiles = await glob("**/*", {
		cwd: args.outDir,
		follow: true,
		includeDirectories: true
	});
	for (const file of Object.keys(compiledSourceFiles)) files[file] = compiledSourceFiles[file];
	/**
	* While we're not using NFT to process source code, we are using it
	* to tree shake node deps, and include any fs reads for files that are
	* not part of the traced paths or compiled source files.
	* Most of this is identical to the `@vercel/node` implementation
	*/
	const runNft = () => nodeFileTrace$1(Array.from(tracedPaths), {
		base: args.repoRootPath,
		processCwd: args.workPath,
		ts: true,
		mixedModules: true,
		async resolve(id, parent, job, cjsResolve) {
			return resolve$1(id, parent, job, cjsResolve);
		},
		async readFile(fsPath) {
			try {
				let source = await readFile(fsPath);
				if (fsPath.endsWith(".ts") && !fsPath.endsWith(".d.ts") || fsPath.endsWith(".tsx") || fsPath.endsWith(".mts") || fsPath.endsWith(".cts")) source = (await transform(fsPath, source.toString())).code;
				return source;
			} catch (error) {
				if (isNativeError(error) && "code" in error && (error.code === "ENOENT" || error.code === "EISDIR")) return null;
				throw error;
			}
		}
	});
	const result = await span.child("vc.builder.backends.nft").trace(runNft);
	if (!args.keepTracedPaths) for (const file of tracedPaths) {
		const relativeFile = relative(args.repoRootPath, file);
		result.fileList.delete(relativeFile);
	}
	debug("NFT traced files count:", result.fileList.size);
	for (const file of result.fileList) {
		const absolutePath = join(args.repoRootPath, file);
		const stats = await lstat(absolutePath);
		const outputPath = file;
		if (stats.isSymbolicLink() || stats.isFile()) files[outputPath] = new FileFsRef({
			fsPath: absolutePath,
			mode: stats.mode
		});
	}
	debug("Total files in context:", Object.keys(files).length);
	return files;
};

//#endregion
//#region src/cervel/rolldown.ts
const __dirname__filenameShim = `
import { createRequire as __createRequire } from 'node:module';
import { fileURLToPath as __fileURLToPath } from 'node:url';
import { dirname as __dirname_ } from 'node:path';
var require = typeof require !== 'undefined' ? require : __createRequire(import.meta.url);
var __filename = typeof __filename !== 'undefined' ? __filename : __fileURLToPath(import.meta.url);
var __dirname = typeof __dirname !== 'undefined' ? __dirname : __dirname_(__filename);
`.trim();
const rolldown$1 = async (args) => {
	const entrypointPath = join(args.workPath, args.entrypoint);
	const outputDir = join(args.workPath, args.out);
	const extension = extname(args.entrypoint);
	const extensionMap = {
		".ts": {
			format: "auto",
			extension: "js"
		},
		".mts": {
			format: "esm",
			extension: "mjs"
		},
		".cts": {
			format: "cjs",
			extension: "cjs"
		},
		".cjs": {
			format: "cjs",
			extension: "cjs"
		},
		".js": {
			format: "auto",
			extension: "js"
		},
		".mjs": {
			format: "esm",
			extension: "mjs"
		}
	};
	const extensionInfo = extensionMap[extension] || extensionMap[".js"];
	let resolvedFormat = extensionInfo.format === "auto" ? void 0 : extensionInfo.format;
	const packageJsonPath = join(args.workPath, "package.json");
	const external = [];
	let pkg = {};
	if (existsSync(packageJsonPath)) {
		const source = await readFile(packageJsonPath, "utf8");
		try {
			pkg = JSON.parse(source.toString());
		} catch (_e) {
			pkg = {};
		}
		if (extensionInfo.format === "auto") if (pkg.type === "module") resolvedFormat = "esm";
		else resolvedFormat = "cjs";
		for (const dependency of Object.keys(pkg.dependencies || {})) external.push(dependency);
		for (const dependency of Object.keys(pkg.devDependencies || {})) external.push(dependency);
		for (const dependency of Object.keys(pkg.peerDependencies || {})) external.push(dependency);
		for (const dependency of Object.keys(pkg.optionalDependencies || {})) external.push(dependency);
	}
	const resolvedExtension = resolvedFormat === "esm" ? "mjs" : "cjs";
	const context = { tracedPaths: /* @__PURE__ */ new Set() };
	const runRolldown = () => build$2({
		input: entrypointPath,
		cwd: args.workPath,
		platform: "node",
		tsconfig: true,
		plugins: [plugin({
			repoRootPath: args.repoRootPath,
			outDir: outputDir,
			workPath: args.workPath,
			shimBareImports: resolvedFormat === "esm",
			context
		})],
		output: {
			cleanDir: true,
			dir: outputDir,
			format: resolvedFormat,
			entryFileNames: `[name].${resolvedExtension}`,
			preserveModules: true,
			preserveModulesRoot: args.repoRootPath,
			sourcemap: false,
			banner: resolvedFormat === "esm" ? __dirname__filenameShim : void 0
		}
	});
	const out = await args.span.child("vc.builder.backends.rolldown").trace(runRolldown);
	let handler = null;
	for (const entry of out.output) if (entry.type === "chunk") {
		if (entry.isEntry) handler = entry.fileName;
	}
	if (typeof handler !== "string") throw new Error(`Unable to resolve module for ${args.entrypoint}`);
	const outputFiles = await nodeFileTrace({
		outDir: outputDir,
		tracedPaths: Array.from(context.tracedPaths),
		repoRootPath: args.repoRootPath,
		workPath: args.workPath,
		keepTracedPaths: false,
		span: args.span
	});
	const cleanup = async () => {
		await rm(outputDir, {
			recursive: true,
			force: true
		});
	};
	return {
		result: {
			handler,
			outputDir,
			outputFiles
		},
		cleanup
	};
};

//#endregion
//#region src/cervel/utils.ts
const noColor = globalThis.process?.env?.NO_COLOR === "1" || globalThis.process?.env?.TERM === "dumb";
const resets = {
	1: 22,
	31: 39,
	32: 39,
	33: 39,
	34: 39,
	35: 39,
	36: 39,
	90: 39
};
const _c = (c) => (text) => {
	if (noColor) return text;
	return `\u001B[${c}m${text}\u001B[${resets[c] ?? 0}m`;
};
const Colors = {
	bold: _c(1),
	red: _c(31),
	green: _c(32),
	yellow: _c(33),
	blue: _c(34),
	magenta: _c(35),
	cyan: _c(36),
	gray: _c(90),
	url: (title, url) => noColor ? `[${title}](${url})` : `\u001B]8;;${url}\u001B\\${title}\u001B]8;;\u001B\\`
};

//#endregion
//#region src/typescript.ts
/**
* Typecheck via the TypeScript compiler API (`createProgram`, `getPreEmitDiagnostics`),
* not by spawning the `tsc` binary.
*
* We only want to validate the deployment entrypoint and its import graph, not every
* file matched by `tsconfig` `include`. The CLI cannot combine `--project` with explicit
* root files (TS5042), so expressing 'project options + entry-only roots' in one `tsc`
* call requires a generated tsconfig on disk. Writing beside the user's config is
* invasive; a temp config elsewhere often breaks `node_modules` / `@types` resolution
* relative to the real project. The API lets us reuse `parseJsonConfigFileContent` (same
* options as `-p`) with explicit `rootNames`, no files written, and a compiler host whose
* current directory stays `workPath`.
*
* The `typescript` package is resolved with `require` from the user's app (peer dependency), not bundled.
*/
const require_ = createRequire(import.meta.url);
const typescript = (args) => {
	const { span } = args;
	return span.child("vc.builder.backends.tsCompile").trace(async () => {
		const extension = extname(args.entrypoint);
		if (![
			".ts",
			".mts",
			".cts"
		].includes(extension)) return;
		const ts = resolveTypeScriptModule(args.workPath);
		if (!ts) {
			console.log(Colors.gray(`${Colors.bold(Colors.cyan("✓"))} Typecheck skipped ${Colors.gray("(TypeScript not found)")}`));
			return null;
		}
		return doTypeCheck(args, ts);
	});
};
async function doTypeCheck(args, ts) {
	const entryAbsolute = resolve(args.workPath, args.entrypoint);
	const tsconfig = await findNearestTsconfig(args.workPath);
	const formatDiagnostics = process.stdout.isTTY ? ts.formatDiagnosticsWithColorAndContext : ts.formatDiagnostics;
	const diagnosticHost = {
		getNewLine: () => ts.sys.newLine,
		getCanonicalFileName: (fileName) => ts.sys.useCaseSensitiveFileNames ? fileName : fileName.toLowerCase(),
		getCurrentDirectory: () => args.workPath
	};
	let options;
	let parseDiagnostics = [];
	if (tsconfig) {
		const configRead = ts.readConfigFile(tsconfig, ts.sys.readFile);
		if (configRead.error) {
			const message = formatDiagnostics([configRead.error], diagnosticHost);
			console.error("\nTypeScript type check failed:\n");
			console.error(message);
			throw new Error("TypeScript type check failed");
		}
		const parsed = ts.parseJsonConfigFileContent(configRead.config, ts.sys, dirname(tsconfig), void 0, tsconfig);
		parseDiagnostics = parsed.errors;
		options = {
			...parsed.options,
			noEmit: true,
			skipLibCheck: true,
			allowJs: true,
			esModuleInterop: true
		};
	} else options = {
		noEmit: true,
		skipLibCheck: true,
		allowJs: true,
		esModuleInterop: true,
		target: ts.ScriptTarget.ES2022,
		module: ts.ModuleKind.NodeNext,
		moduleResolution: ts.ModuleResolutionKind.NodeNext
	};
	const compilerHost = ts.createCompilerHost(options);
	compilerHost.getCurrentDirectory = () => args.workPath;
	const program = ts.createProgram([entryAbsolute], options, compilerHost);
	const errors = [...parseDiagnostics, ...ts.getPreEmitDiagnostics(program)].filter((d) => d.category === ts.DiagnosticCategory.Error);
	if (errors.length === 0) {
		console.log(Colors.gray(`${Colors.bold(Colors.cyan("✓"))} Typecheck complete`));
		return;
	}
	const output = formatDiagnostics(errors, diagnosticHost);
	console.error("\nTypeScript type check failed:\n");
	console.error(output);
	throw new Error("TypeScript type check failed");
}
function resolveTypeScriptModule(workPath) {
	try {
		return require_(require_.resolve("typescript", { paths: [workPath] }));
	} catch (_e) {
		return null;
	}
}
const findNearestTsconfig = async (workPath) => {
	const tsconfigPath = join(workPath, "tsconfig.json");
	if (existsSync(tsconfigPath)) return tsconfigPath;
	if (workPath === "/") return;
	return findNearestTsconfig(join(workPath, ".."));
};

//#endregion
//#region src/find-entrypoint.ts
const frameworks = [
	"express",
	"hono",
	"elysia",
	"fastify",
	"@nestjs/core",
	"h3"
];
const entrypointFilenames = [
	"app",
	"index",
	"server",
	"main",
	"src/app",
	"src/index",
	"src/server",
	"src/main"
];
const entrypointExtensions = [
	"js",
	"cjs",
	"mjs",
	"ts",
	"cts",
	"mts"
];
const entrypoints = entrypointFilenames.flatMap((filename) => entrypointExtensions.map((extension) => `${filename}.${extension}`));
const createFrameworkRegex = (framework) => new RegExp(`(?:from|require|import)\\s*(?:\\(\\s*)?["']${framework}["']\\s*(?:\\))?`, "g");
const findEntrypoint = async (cwd) => {
	let packageJsonObject = null;
	try {
		const packageJson = await readFile(join(cwd, "package.json"), "utf-8");
		packageJsonObject = JSON.parse(packageJson);
	} catch (_) {}
	if (packageJsonObject) {
		const main = typeof packageJsonObject.main === "string" ? packageJsonObject.main.trim() : "";
		if (main) {
			const abs = resolve(cwd, main);
			const rel = relative(cwd, abs);
			if (!rel.startsWith("..") && rel !== "") try {
				await readFile(abs, "utf-8");
				return rel.split(sep).join("/");
			} catch {}
		}
	}
	let framework;
	if (packageJsonObject) framework = frameworks.find((framework$1) => packageJsonObject.dependencies?.[framework$1]);
	if (!framework) for (const entrypoint of entrypoints) {
		const entrypointPath = join(cwd, entrypoint);
		try {
			await readFile(entrypointPath, "utf-8");
			return entrypoint;
		} catch (_) {}
	}
	const regex = framework ? createFrameworkRegex(framework) : void 0;
	for (const entrypoint of entrypoints) {
		const entrypointPath = join(cwd, entrypoint);
		try {
			const content = await readFile(entrypointPath, "utf-8");
			if (regex) {
				if (regex.test(content)) return entrypoint;
			}
		} catch (_) {}
	}
};
const findEntrypointOrThrow = async (cwd) => {
	const entrypoint = await findEntrypoint(cwd);
	if (!entrypoint) throw new Error(`No entrypoint found in "${cwd}". Set package.json "main" to a server file, or add one of: ${entrypoints.join(", ")}`);
	return entrypoint;
};

//#endregion
//#region src/cervel/index.ts
const require$2 = createRequire(import.meta.url);
const getBuildSummary = async (outputDir) => {
	const buildSummary = await readFile$1(join(outputDir, ".cervel.json"), "utf-8");
	return JSON.parse(buildSummary);
};
const build$1 = async (args) => {
	const entrypoint = args.entrypoint || await findEntrypointOrThrow(args.workPath);
	const span = args.span ?? new Span({ name: "cervel-build" });
	const [, rolldownResult] = await Promise.all([typescript({
		entrypoint,
		workPath: args.workPath,
		span
	}), rolldown$1({
		entrypoint,
		workPath: args.workPath,
		repoRootPath: args.repoRootPath,
		out: args.out,
		span
	})]);
	await writeFile(join(args.workPath, args.out, ".cervel.json"), JSON.stringify({ handler: rolldownResult.result.handler }, null, 2));
	console.log(Colors.gray(`${Colors.bold(Colors.cyan("✓"))} Build complete — Using ${Colors.bold(entrypoint)} as the root entrypoint.`));
	return { rolldownResult: rolldownResult.result };
};
const serve = async (args) => {
	const entrypoint = await findEntrypointOrThrow(args.workPath);
	const srvxBin = join(require$2.resolve("srvx"), "..", "..", "..", "bin", "srvx.mjs");
	const tsxBin = require$2.resolve("tsx");
	const restArgs = Object.entries(args.rest).filter(([, value]) => value !== void 0 && value !== false).map(([key, value]) => typeof value === "boolean" ? `--${key}` : `--${key}=${value}`);
	if (!args.rest.import) restArgs.push("--import", tsxBin);
	await execa("npx", [
		srvxBin,
		...restArgs,
		entrypoint
	], {
		cwd: args.workPath,
		stdio: "inherit"
	});
};
const srvxOptions = {
	help: {
		type: "boolean",
		short: "h"
	},
	version: {
		type: "boolean",
		short: "v"
	},
	prod: { type: "boolean" },
	port: {
		type: "string",
		short: "p"
	},
	host: {
		type: "string",
		short: "H"
	},
	static: {
		type: "string",
		short: "s"
	},
	import: { type: "string" },
	tls: { type: "boolean" },
	cert: { type: "string" },
	key: { type: "string" }
};

//#endregion
//#region src/rolldown/resolve-format.ts
const resolveEntrypointAndFormat = async (args) => {
	const extension = extname(args.entrypoint);
	const extensionMap = {
		".ts": {
			format: "auto",
			extension: "js"
		},
		".mts": {
			format: "esm",
			extension: "mjs"
		},
		".cts": {
			format: "cjs",
			extension: "cjs"
		},
		".cjs": {
			format: "cjs",
			extension: "cjs"
		},
		".js": {
			format: "auto",
			extension: "js"
		},
		".mjs": {
			format: "esm",
			extension: "mjs"
		}
	};
	const extensionInfo = extensionMap[extension] || extensionMap[".js"];
	let resolvedFormat = extensionInfo.format === "auto" ? args.defaultFormat : extensionInfo.format;
	const packageJsonPath = join(args.workPath, "package.json");
	let pkg = {};
	if (existsSync(packageJsonPath)) {
		const source = await readFile(packageJsonPath, "utf8");
		try {
			pkg = JSON.parse(source.toString());
		} catch (_e) {
			pkg = {};
		}
		if (extensionInfo.format === "auto") if (pkg.type === "module") resolvedFormat = "esm";
		else resolvedFormat = "cjs";
	}
	if (!resolvedFormat) throw new Error(`Unable to resolve format for ${args.entrypoint}`);
	return {
		format: resolvedFormat,
		extension: resolvedFormat === "esm" ? "mjs" : "cjs"
	};
};

//#endregion
//#region src/service-vc-init.ts
async function applyServiceVcInit(args) {
	const { format, extension } = await resolveShimFormat$1(args);
	const handlerDir = dirname(args.handler);
	const vcInitName = `${basename(args.handler, extname(args.handler))}.__vc_service_vc_init${extension}`;
	const vcInitHandler = handlerDir === "." ? vcInitName : join(handlerDir, vcInitName);
	const handlerImportPath = `./${basename(args.handler)}`;
	const vcInitSource = format === "esm" ? createEsmServiceVcInit(handlerImportPath) : createCjsServiceVcInit(handlerImportPath);
	return {
		handler: vcInitHandler,
		files: {
			...args.files,
			[vcInitHandler]: new FileBlob({
				data: vcInitSource,
				mode: 420
			})
		}
	};
}
async function resolveShimFormat$1(args) {
	const { format } = await resolveEntrypointAndFormat({
		entrypoint: args.handler,
		workPath: args.workPath
	});
	return {
		format,
		extension: extname(args.handler) || (format === "esm" ? ".mjs" : ".cjs")
	};
}
const sharedShimPrelude = String.raw`
const PATCH_SYMBOL = Symbol.for('vc.service.route-prefix-strip.patch')

function normalizeServiceRoutePrefix(rawPrefix) {
  if (!rawPrefix) {
    return ''
  }

  let prefix = String(rawPrefix).trim()
  if (!prefix) {
    return ''
  }

  if (!prefix.startsWith('/')) {
    prefix = '/' + prefix
  }

  if (prefix !== '/') {
    prefix = prefix.replace(/\/+$/, '')
  }

  return prefix === '/' ? '' : prefix
}

function getServiceRoutePrefix() {
  const enabled = String(
    process.env.VERCEL_SERVICE_ROUTE_PREFIX_STRIP || ''
  ).toLowerCase()
  if (enabled !== '1' && enabled !== 'true') {
    return ''
  }

  return normalizeServiceRoutePrefix(process.env.VERCEL_SERVICE_ROUTE_PREFIX || '')
}

function stripServiceRoutePrefix(requestUrl, prefix) {
  if (typeof requestUrl !== 'string' || requestUrl === '*') {
    return requestUrl
  }

  const queryIndex = requestUrl.indexOf('?')
  const rawPath =
    queryIndex === -1 ? requestUrl : requestUrl.slice(0, queryIndex)
  const query = queryIndex === -1 ? '' : requestUrl.slice(queryIndex)

  let path = rawPath || '/'
  if (!path.startsWith('/')) {
    path = '/' + path
  }

  if (!prefix) {
    return path + query
  }

  if (path === prefix) {
    return '/' + query
  }

  if (path.startsWith(prefix + '/')) {
    return path.slice(prefix.length) + query
  }

  return path + query
}

function patchServerRequestUrl(ServerCtor) {
  const prefix = getServiceRoutePrefix()
  if (!prefix || globalThis[PATCH_SYMBOL]) {
    return
  }

  globalThis[PATCH_SYMBOL] = true

  const originalEmit = ServerCtor.prototype.emit
  ServerCtor.prototype.emit = function patchedEmit(event, request, ...args) {
    if (event === 'request' && request && typeof request.url === 'string') {
      request.url = stripServiceRoutePrefix(request.url, prefix)
    }

    return originalEmit.call(this, event, request, ...args)
  }
}
`;
function createEsmServiceVcInit(handlerImportPath) {
	return `
import { Server } from 'node:http'

${sharedShimPrelude}

// Patch the HTTP server before loading user code so apps that attach request
// listeners during module evaluation see the stripped service-relative URL.
patchServerRequestUrl(Server)

const originalModule = await import(${JSON.stringify(handlerImportPath)})

/**
 * Match the Node serverless loader behavior: TS/CJS/ESM interop can leave us
 * with nested \`.default\` wrappers, so peel off a few layers to recover the
 * actual user entrypoint shape.
 */
function unwrapDefaultExport(value) {
  let current = value
  for (let i = 0; i < 5; i++) {
    if (current && typeof current === 'object' && 'default' in current && current.default) {
      current = current.default
    } else {
      break
    }
  }
  return current
}

const entrypoint = unwrapDefaultExport(originalModule)

// Re-export the resolved entrypoint so the surrounding runtime still sees the
// same handler shape after this service bootstrap runs.
export default typeof entrypoint === 'undefined' ? originalModule : entrypoint
`.trimStart();
}
function createCjsServiceVcInit(handlerImportPath) {
	return `
const { Server } = require('node:http')

${sharedShimPrelude}

patchServerRequestUrl(Server)

module.exports = require(${JSON.stringify(handlerImportPath)})
`.trimStart();
}

//#endregion
//#region src/cron-dispatch.ts
const TEMPLATES_DIR = join(dirname(fileURLToPath(import.meta.url)), "..", "templates");
const ESM_TEMPLATE = readFileSync(join(TEMPLATES_DIR, "vc_cron_dispatch.mjs"), "utf-8");
const CJS_TEMPLATE = readFileSync(join(TEMPLATES_DIR, "vc_cron_dispatch.cjs"), "utf-8");
const USER_MODULE_PLACEHOLDER = /['"]__VC_USER_MODULE_PATH__['"]/g;
const ROUTES_PLACEHOLDER = /'__VC_ROUTES_JSON__'/g;
/**
* Wrap a cron service handler with a dispatcher shim that:
*   - looks up the inbound request path in a route table baked into the
*     shim at build time and invokes the named export on the user module
*     (or the default export when the table value is `"default"`)
*   - verifies `CRON_SECRET` via `Authorization: Bearer ...` when set
*
* The dispatcher source lives in templates/vc_cron_dispatch.{mjs,cjs};
* this function picks the right template, swaps in the user module
* import path and the cron route table, and writes the result into the
* lambda files.
*
* The route table is embedded inline rather than passed via a lambda
* env var because AWS Lambda rejects env var names that don't start
* with a letter — `__VC_CRON_ROUTES` would fail at deploy time. The
* Python builder works around the same constraint by writing the route
* table into its trampoline source.
*/
async function applyCronDispatch(args) {
	const { format, extension } = await resolveShimFormat(args);
	const handlerDir = posix.dirname(args.handler);
	const dispatchName = `${posix.basename(args.handler, extname(args.handler))}.__vc_cron_dispatch${extension}`;
	const dispatchHandler = handlerDir === "." ? dispatchName : posix.join(handlerDir, dispatchName);
	const handlerImportPath = `./${posix.basename(args.handler)}`;
	const routesJson = JSON.stringify(args.routes);
	if (routesJson.includes("\\") || routesJson.includes("'")) throw new Error(`cron route table contains characters that need JS-string escaping: ${routesJson}`);
	const dispatchSource = (format === "esm" ? ESM_TEMPLATE : CJS_TEMPLATE).replace(USER_MODULE_PLACEHOLDER, JSON.stringify(handlerImportPath)).replace(ROUTES_PLACEHOLDER, `'${routesJson}'`);
	return {
		handler: dispatchHandler,
		files: {
			...args.files,
			[dispatchHandler]: new FileBlob({
				data: dispatchSource,
				mode: 420
			})
		}
	};
}
async function resolveShimFormat(args) {
	const { format } = await resolveEntrypointAndFormat({
		entrypoint: args.handler,
		workPath: args.workPath,
		defaultFormat: "cjs"
	});
	return {
		format,
		extension: extname(args.handler) || (format === "esm" ? ".mjs" : ".cjs")
	};
}

//#endregion
//#region src/crons.ts
const DYNAMIC_SCHEDULE = "<dynamic>";
/** Build the JSON route table embedded in `__VC_CRON_ROUTES`. */
function buildCronRouteTable(crons) {
	const table = {};
	for (const cron of crons) table[cron.path] = "default";
	return table;
}
/**
* Compute cron entries for a JS/TS cron service build.
*
* Mirrors `packages/python/src/crons.ts` for static schedules. Returns
* `undefined` when the service is not schedule-triggered. Throws on
* `<dynamic>` schedules — that path is reserved for a follow-up.
*
* v1 always invokes the user module's default export, so this returns
* plain `Cron[]` (no handler-name field). When `handlerFunction` or
* `<dynamic>` support lands, this will need to grow a per-path handler
* name back.
*/
function getServiceCrons(opts) {
	const { service, entrypoint } = opts;
	if (!service || !isScheduleTriggeredService(service)) return;
	if (!service.name || typeof service.schedule !== "string") return;
	if (!entrypoint) throw new Error("Cron service is missing an entrypoint");
	if (service.schedule === DYNAMIC_SCHEDULE) throw new Error("Dynamic cron schedules (\"<dynamic>\") are not yet supported for JavaScript/TypeScript services. Use a static cron expression in vercel.json.");
	return [{
		path: getInternalServiceCronPath(service.name, entrypoint, "cron"),
		schedule: service.schedule
	}];
}

//#endregion
//#region src/rolldown/nft.ts
const nft = async (args) => {
	const nftSpan = args.span.child("vc.builder.backends.nft");
	const runNft = async () => {
		const ignorePatterns = [...args.ignoreNodeModules ? ["**/node_modules/**"] : [], ...args.ignore ? Array.isArray(args.ignore) ? args.ignore : [args.ignore] : []];
		const nftResult = await nodeFileTrace$1(Array.from(args.localBuildFiles).filter((p) => existsSync(p)), {
			base: args.repoRootPath,
			processCwd: args.workPath,
			ts: true,
			mixedModules: true,
			conditions: args.conditions,
			ignore: ignorePatterns.length > 0 ? ignorePatterns : void 0,
			async readFile(fsPath) {
				try {
					let source = await readFile(fsPath);
					if (isTypeScriptFile(fsPath)) source = (await transform(fsPath, source.toString())).code;
					return source;
				} catch (error) {
					if (isNativeError(error) && "code" in error && (error.code === "ENOENT" || error.code === "EISDIR")) return null;
					throw error;
				}
			}
		});
		for (const file of nftResult.fileList) {
			const absolutePath = join(args.repoRootPath, file);
			let stats;
			try {
				stats = await lstat(absolutePath);
			} catch (error) {
				if (isNativeError(error) && "code" in error && error.code === "ENOENT") continue;
				throw error;
			}
			const outputPath = file;
			if (args.localBuildFiles.has(join(args.repoRootPath, outputPath))) continue;
			if (stats.isSymbolicLink() || stats.isFile()) if (args.ignoreNodeModules) {
				if ((stats.isSymbolicLink() ? await stat(absolutePath) : stats).isFile()) {
					const content = await readFile(absolutePath, "utf-8");
					args.files[outputPath] = new FileBlob({
						data: content,
						mode: stats.mode
					});
				}
			} else args.files[outputPath] = new FileFsRef({
				fsPath: absolutePath,
				mode: stats.mode
			});
		}
	};
	await nftSpan.trace(runNft);
};
const isTypeScriptFile = (fsPath) => {
	if (fsPath.endsWith(".d.ts") || fsPath.endsWith(".d.mts") || fsPath.endsWith(".d.cts")) return false;
	return fsPath.endsWith(".ts") || fsPath.endsWith(".tsx") || fsPath.endsWith(".mts") || fsPath.endsWith(".cts");
};

//#endregion
//#region src/rolldown/index.ts
const PLUGIN_NAME = "vercel:backends";
const CJS_SHIM_PREFIX = "\0cjs-shim:";
const rolldown = async (args) => {
	const files = {};
	const { format, extension } = await resolveEntrypointAndFormat(args);
	const localBuildFiles = /* @__PURE__ */ new Set();
	let handler = null;
	const packageJsonCache = /* @__PURE__ */ new Map();
	const shimMeta = /* @__PURE__ */ new Map();
	const framework = {
		slug: "",
		version: ""
	};
	const getPackageJson = async (pkgPath) => {
		if (packageJsonCache.has(pkgPath)) return packageJsonCache.get(pkgPath);
		try {
			const contents = await readFile(pkgPath, "utf-8");
			const parsed = JSON.parse(contents);
			packageJsonCache.set(pkgPath, parsed);
			return parsed;
		} catch {
			packageJsonCache.set(pkgPath, null);
			return null;
		}
	};
	const isCommonJS = async (bareImport, resolvedPath, resolvedInfo) => {
		const ext = extname(resolvedPath);
		if (ext === ".cjs") return true;
		if (ext === ".mjs") return false;
		if (ext === ".js" || ext === ".ts") {
			const pkgJsonPath = resolvedInfo.packageJsonPath;
			if (!pkgJsonPath) return true;
			const pkgJson = await getPackageJson(pkgJsonPath);
			if (!pkgJson) return true;
			const pkgDir = dirname(pkgJsonPath);
			const relativePath = resolvedPath.slice(pkgDir.length + 1).replace(/\\/g, "/");
			const pkgName = pkgJson.name || "";
			const subpath = bareImport.startsWith(pkgName) ? `.${bareImport.slice(pkgName.length)}` || "." : ".";
			try {
				if (exports(pkgJson, subpath, {
					require: false,
					conditions: ["node", "import"]
				})?.some((p) => p === relativePath || p === `./${relativePath}`)) return false;
				if (exports(pkgJson, subpath, {
					require: true,
					conditions: ["node", "require"]
				})?.some((p) => p === relativePath || p === `./${relativePath}`)) return true;
			} catch {}
			if (pkgJson.module) return false;
			return pkgJson.type !== "module";
		}
		return true;
	};
	const isBareImport = (id) => {
		return !id.startsWith(".") && !id.startsWith("/") && !/^[a-z][a-z0-9+.-]*:/i.test(id);
	};
	const isNodeModule = (resolved) => {
		return resolved?.id?.includes("node_modules") ?? false;
	};
	const isNodeBuiltin = (id) => {
		const normalizedId = id.includes(":") ? id.split(":")[1] : id;
		return builtinModules.includes(normalizedId);
	};
	const isBunBuiltin = (id) => {
		return id === "bun" || id.startsWith("bun:");
	};
	const isLocalImport = (id) => {
		return !id.startsWith("node:") && !id.includes("node_modules");
	};
	const plugin$1 = () => {
		return {
			name: PLUGIN_NAME,
			resolveId: {
				order: "pre",
				async handler(id, importer, rOpts) {
					if (id.startsWith(CJS_SHIM_PREFIX)) return {
						id,
						external: false
					};
					const resolved = await this.resolve(id, importer, rOpts);
					if (isNodeBuiltin(id)) return {
						id: id.startsWith("node:") ? id : `node:${id}`,
						external: true
					};
					if (isBunBuiltin(id)) return {
						id,
						external: true
					};
					if (resolved?.id && isLocalImport(resolved.id)) {
						if (existsSync(resolved.id)) localBuildFiles.add(resolved.id);
					} else if (!resolved && !(isBareImport(id) && importer)) {
						const candidate = join(args.workPath, id);
						if (existsSync(candidate)) localBuildFiles.add(candidate);
					}
					if (importer?.startsWith(CJS_SHIM_PREFIX) && isBareImport(id)) return {
						id,
						external: true
					};
					if (importer && isBareImport(id) && isNodeModule(resolved)) {
						if (isBackendFramework(id) && resolved?.packageJsonPath) try {
							const pkg = await readFile(resolved.packageJsonPath, "utf8");
							const pkgJson = JSON.parse(pkg);
							framework.slug = pkgJson.name;
							framework.version = pkgJson.version;
						} catch {}
						if (resolved ? await isCommonJS(id, resolved.id, resolved) : false) {
							const importerPkgJsonPath = (await this.resolve(importer))?.packageJsonPath;
							if (importerPkgJsonPath) {
								const importerPkgDir = relative(args.repoRootPath, dirname(importerPkgJsonPath));
								const shimId$1 = `${CJS_SHIM_PREFIX}${importerPkgDir.replace(/\//g, "_")}_${id.replace(/\//g, "_")}`;
								shimMeta.set(shimId$1, {
									pkgDir: importerPkgDir,
									pkgName: id
								});
								return {
									id: shimId$1,
									external: false
								};
							}
							const shimId = `${CJS_SHIM_PREFIX}${id.replace(/\//g, "_")}`;
							shimMeta.set(shimId, {
								pkgDir: "",
								pkgName: id
							});
							return {
								id: shimId,
								external: false
							};
						}
						return {
							id,
							external: true
						};
					}
					if (importer && isBareImport(id)) return resolved;
					if (resolved && !isNodeModule(resolved)) return resolved;
					return resolved;
				}
			},
			load: { async handler(id) {
				if (id.startsWith(CJS_SHIM_PREFIX)) {
					const meta = shimMeta.get(id);
					if (!meta) return { code: `module.exports = require('${id.slice(10)}');` };
					const { pkgDir, pkgName } = meta;
					return { code: `
import { createRequire } from 'node:module';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

const requireFromContext = createRequire(join(dirname(fileURLToPath(import.meta.url)), '${pkgDir ? join("..", pkgDir, "package.json") : "../package.json"}'));
module.exports = requireFromContext('${pkgName}');
`.trim() };
				}
				return null;
			} }
		};
	};
	const runRolldown = () => build$2({
		input: args.entrypoint,
		write: false,
		cwd: args.workPath,
		platform: "node",
		transform: { define: format === "esm" ? {
			__dirname: "import.meta.dirname",
			__filename: "import.meta.filename"
		} : void 0 },
		tsconfig: true,
		plugins: [plugin$1()],
		output: {
			cleanDir: true,
			format,
			entryFileNames: `[name].${extension}`,
			preserveModules: true,
			preserveModulesRoot: args.repoRootPath,
			sourcemap: false
		}
	});
	const rolldownSpan = args.span?.child("vc.builder.backends.rolldown");
	const out = await rolldownSpan?.trace(runRolldown) || await runRolldown();
	for (const file of out.output) if (file.type === "chunk") {
		if (file.isEntry) handler = file.fileName;
		files[file.fileName] = new FileBlob({
			data: file.code,
			mode: 420
		});
	}
	await nft({
		...args,
		localBuildFiles,
		files,
		span: rolldownSpan ?? new Span({ name: "vc.builder.backends.nft" }),
		ignoreNodeModules: true,
		ignore: args.config.excludeFiles
	});
	if (!handler) throw new Error(`Unable to resolve build handler for entrypoint: ${args.entrypoint}`);
	console.log(Colors.gray(`${Colors.bold(Colors.cyan("✓"))} Build complete — Using ${Colors.bold(args.entrypoint)} as the root entrypoint.`));
	return {
		files,
		handler,
		framework,
		localBuildFiles
	};
};

//#endregion
//#region src/rolldown/util.ts
const BEGIN_INTROSPECTION_RESULT = "\n__VERCEL_INTROSPECTION_BEGIN__\n";
const END_INTROSPECTION_RESULT = "\n__VERCEL_INTROSPECTION_END__\n";

//#endregion
//#region src/rolldown/introspection.ts
const require$1 = createRequire(import.meta.url);
const introspectionSchema = z.object({
	routes: z.array(z.object({
		src: z.string(),
		dest: z.string(),
		methods: z.array(z.string())
	})),
	additionalFolders: z.array(z.string()).optional(),
	additionalDeps: z.array(z.string()).optional()
});
const introspection = async (args) => {
	const defaultResult$1 = {
		routes: [],
		additionalFolders: [],
		additionalDeps: []
	};
	if (isExperimentalBackendsWithoutIntrospectionEnabled()) return defaultResult$1;
	const introspectionSpan = args.span.child("vc.builder.backends.introspection");
	const runIntrospection = async () => {
		const rolldownEsmLoaderPath = `file://${require$1.resolve("@vercel/backends/rolldown/esm")}`;
		const rolldownCjsLoaderPath = require$1.resolve("@vercel/backends/rolldown/cjs-hooks");
		const handlerPath = join(args.workPath, args.entrypoint);
		const files = args.files;
		const tmpDir = mkdtempSync(join(tmpdir(), "vercel-introspection-"));
		for (const [key, value] of Object.entries(files)) {
			if (!(value instanceof FileBlob) || typeof value.data !== "string") continue;
			const filePath = join(tmpDir, key);
			mkdirSync(dirname(filePath), { recursive: true });
			writeFileSync(filePath, value.data);
		}
		let introspectionData;
		let introspectionError;
		await new Promise((resolvePromise) => {
			try {
				debug("Spawning introspection process");
				const outputTempDir = mkdtempSync(join(tmpdir(), "introspection-output-"));
				const tempFilePath = join(outputTempDir, "output.txt");
				const writeStream = createWriteStream(tempFilePath);
				let streamClosed = false;
				const child = spawn("node", [
					"-r",
					rolldownCjsLoaderPath,
					"--import",
					rolldownEsmLoaderPath,
					handlerPath
				], {
					stdio: [
						"pipe",
						"pipe",
						"pipe"
					],
					cwd: args.workPath,
					env: {
						...process.env,
						...args.meta?.buildEnv,
						...args.meta?.env,
						VERCEL_INTROSPECTION_HANDLER: handlerPath,
						VERCEL_INTROSPECTION_HANDLER_BUILT: args.handler,
						VERCEL_INTROSPECTION_WORK_PATH: args.workPath,
						VERCEL_INTROSPECTION_REPO_ROOT_PATH: args.repoRootPath,
						VERCEL_INTROSPECTION_TMP_DIR: tmpDir
					}
				});
				child.stdout?.pipe(writeStream);
				let stderrBuffer = "";
				child.stderr?.on("data", (data) => {
					stderrBuffer += data.toString();
				});
				writeStream.on("error", (err) => {
					debug(`Write stream error: ${err.message}`);
				});
				const timeout = setTimeout(() => {
					debug("Introspection timeout, killing process with SIGTERM");
					child.kill("SIGTERM");
				}, 5e3);
				const timeout2 = setTimeout(() => {
					debug("Introspection timeout, killing process with SIGKILL");
					child.kill("SIGKILL");
				}, 6e3);
				const cleanup = () => {
					clearTimeout(timeout);
					clearTimeout(timeout2);
					try {
						rmSync(tmpDir, {
							recursive: true,
							force: true
						});
					} catch (err) {
						debug(`Error deleting tmpDir: ${err}`);
					}
				};
				child.on("error", (err) => {
					cleanup();
					introspectionError = `Loader error: ${err.message}`;
					debug(introspectionError);
					if (!streamClosed) writeStream.end(() => {
						streamClosed = true;
						try {
							rmSync(outputTempDir, {
								recursive: true,
								force: true
							});
						} catch (cleanupErr) {
							debug(`Error deleting output temp dir: ${cleanupErr}`);
						}
						resolvePromise();
					});
					else resolvePromise();
				});
				child.on("close", () => {
					cleanup();
					debug("Introspection process closed");
					if (!streamClosed) writeStream.end(() => {
						streamClosed = true;
						let stdoutBuffer;
						try {
							stdoutBuffer = readFileSync(tempFilePath, "utf8");
							const beginIndex = stdoutBuffer.indexOf(BEGIN_INTROSPECTION_RESULT);
							const endIndex = stdoutBuffer.indexOf(END_INTROSPECTION_RESULT);
							if (beginIndex !== -1 && endIndex !== -1) {
								const introspectionString = stdoutBuffer.substring(beginIndex + BEGIN_INTROSPECTION_RESULT.length, endIndex);
								if (introspectionString) {
									introspectionData = introspectionSchema.parse(JSON.parse(introspectionString));
									debug("Introspection data parsed successfully");
								}
							} else {
								introspectionError = `Introspection markers not found. stderr: ${stderrBuffer}`;
								debug(`Introspection markers not found.\nstdout:\n${stdoutBuffer}\nstderr:\n${stderrBuffer}`);
							}
						} catch (error) {
							introspectionError = `Error parsing introspection data: ${error}. stderr: ${stderrBuffer}`;
							debug(`Error parsing introspection data: ${error}\nstdout:\n${stdoutBuffer}\nstderr:\n${stderrBuffer}`);
						} finally {
							try {
								rmSync(outputTempDir, {
									recursive: true,
									force: true
								});
							} catch (err) {
								debug(`Error deleting output temp directory: ${err}`);
							}
							resolvePromise();
						}
					});
					else resolvePromise();
				});
			} catch (error) {
				introspectionError = `Introspection error: ${error}`;
				debug("Introspection error", error);
				resolvePromise();
			}
		});
		if (!introspectionData) {
			introspectionSpan.setAttributes({
				"introspection.success": "false",
				"introspection.routes": "0",
				...introspectionError && { "introspection.error": introspectionError.slice(0, 1024) }
			});
			return defaultResult$1;
		}
		const additionalFolders = (introspectionData.additionalFolders ?? []).map((val) => {
			if (isAbsolute(val)) return relative(args.workPath, val);
			return val;
		});
		introspectionSpan.setAttributes({
			"introspection.success": "true",
			"introspection.routes": String(introspectionData.routes.length)
		});
		return {
			routes: introspectionData.routes,
			additionalFolders,
			additionalDeps: introspectionData.additionalDeps ?? []
		};
	};
	return introspectionSpan.trace(runIntrospection);
};

//#endregion
//#region src/build.ts
/**
* `outputDirectory` is usually the same project setting the static builder uses
* (Vite/Webpack/etc. client output). We only reuse it for the Node lambda when we
* find a known server entry file under that folder; otherwise we bundle from
* source with rolldown. Errors here are swallowed so static-only output trees do
* not fail the build.
*/
async function findEntrypointInOutputDir(dir) {
	try {
		return await findEntrypoint(dir);
	} catch {
		return;
	}
}
const maybeDoBuildCommand = async (args, downloadResult) => {
	const buildCommandResult = await maybeExecBuildCommand(args, downloadResult);
	const outputSetting = args.config.outputDirectory;
	let outputDir;
	let entrypoint;
	if (buildCommandResult && outputSetting) if (outputSetting) {
		const _outputDir = join(args.workPath, outputSetting);
		const _entrypoint = await findEntrypointInOutputDir(_outputDir);
		if (_entrypoint) {
			outputDir = _outputDir;
			entrypoint = _entrypoint;
		}
	} else for (const outputDirectory of [
		"dist",
		"build",
		"output"
	]) {
		const _outputDir = join(args.workPath, outputDirectory);
		if (existsSync(_outputDir)) {
			const _entrypoint = await findEntrypointInOutputDir(_outputDir);
			if (_entrypoint) {
				outputDir = _outputDir;
				entrypoint = _entrypoint;
				break;
			}
		}
	}
	const localBuildFiles = /* @__PURE__ */ new Set();
	let files;
	if (outputDir && entrypoint) {
		files = await glob("**", outputDir);
		for (const file of Object.keys(files)) localBuildFiles.add(join(outputDir, file));
	}
	return {
		localBuildFiles,
		files,
		handler: entrypoint,
		outputDir
	};
};

//#endregion
//#region src/introspection/index.ts
const require = createRequire(import.meta.url);
const introspectApp = async (args) => {
	const { span } = args;
	const introspectionSpan = span.child("vc.builder.backends.introspection");
	if (isExperimentalBackendsWithoutIntrospectionEnabled()) return defaultResult(args);
	const cjsLoaderPath = require.resolve("@vercel/backends/introspection/loaders/cjs");
	const rolldownEsmLoaderPath = `file://${require.resolve("@vercel/backends/introspection/loaders/rolldown-esm")}`;
	const handlerPath = join(args.dir, args.handler);
	const introspectionSchema$1 = z.object({
		frameworkSlug: z.string().optional(),
		routes: z.array(z.object({
			src: z.string(),
			dest: z.string(),
			methods: z.array(z.string())
		})),
		additionalFolders: z.array(z.string()).optional().transform((values) => {
			return values?.map((val) => {
				if (isAbsolute(val)) return relative(args.dir, val);
				return val;
			});
		}),
		additionalDeps: z.array(z.string()).optional()
	});
	let introspectionData;
	await new Promise((resolvePromise) => {
		try {
			debug("Spawning introspection process");
			const child = spawn("node", [
				"-r",
				cjsLoaderPath,
				"--import",
				rolldownEsmLoaderPath,
				handlerPath
			], {
				stdio: [
					"pipe",
					"pipe",
					"pipe"
				],
				cwd: args.dir,
				env: {
					...process.env,
					...args.env
				}
			});
			const tempDir = mkdtempSync(join(tmpdir(), "introspection-"));
			const tempFilePath = join(tempDir, "output.txt");
			const writeStream = createWriteStream(tempFilePath);
			let streamClosed = false;
			child.stdout?.pipe(writeStream);
			let stderrBuffer = "";
			child.stderr?.on("data", (data) => {
				stderrBuffer += data.toString();
			});
			writeStream.on("error", (err) => {
				debug(`Write stream error: ${err.message}`);
			});
			const timeout = setTimeout(() => {
				debug("Introspection timeout, killing process with SIGTERM");
				child.kill("SIGTERM");
			}, 8e3);
			const timeout2 = setTimeout(() => {
				debug("Introspection timeout, killing process with SIGKILL");
				child.kill("SIGKILL");
			}, 9e3);
			child.on("error", (err) => {
				clearTimeout(timeout);
				clearTimeout(timeout2);
				debug(`Loader error: ${err.message}`);
				if (!streamClosed) writeStream.end(() => {
					streamClosed = true;
					try {
						unlinkSync(tempFilePath);
					} catch (cleanupErr) {
						debug(`Error deleting temp file on error: ${cleanupErr}`);
					}
					resolvePromise(void 0);
				});
				else resolvePromise(void 0);
			});
			child.on("close", () => {
				clearTimeout(timeout);
				clearTimeout(timeout2);
				debug("Introspection process closed");
				if (!streamClosed) writeStream.end(() => {
					streamClosed = true;
					let stdoutBuffer;
					try {
						stdoutBuffer = readFileSync(tempFilePath, "utf8");
						const beginIndex = stdoutBuffer.indexOf(BEGIN_INTROSPECTION_RESULT);
						const endIndex = stdoutBuffer.indexOf(END_INTROSPECTION_RESULT);
						if (beginIndex !== -1 && endIndex !== -1) {
							const introspectionString = stdoutBuffer.substring(beginIndex + BEGIN_INTROSPECTION_RESULT.length, endIndex);
							if (introspectionString) {
								introspectionData = introspectionSchema$1.parse(JSON.parse(introspectionString));
								debug("Introspection data parsed successfully");
							}
						} else debug(`Introspection markers not found.\nstdout:\n${stdoutBuffer}\nstderr:\n${stderrBuffer}`);
					} catch (error) {
						debug(`Error parsing introspection data: ${error}\nstdout:\n${stdoutBuffer}\nstderr:\n${stderrBuffer}`);
					} finally {
						try {
							rmSync(tempDir, {
								recursive: true,
								force: true
							});
						} catch (err) {
							debug(`Error deleting temp directory: ${err}`);
						}
						resolvePromise(void 0);
					}
				});
				else resolvePromise(void 0);
			});
		} catch (error) {
			debug("Introspection error", error);
			resolvePromise(void 0);
		}
	});
	const framework = getFramework(args);
	if (!introspectionData) {
		introspectionSpan.setAttributes({
			"introspection.success": "false",
			"introspection.routes": "0"
		});
		return defaultResult(args);
	}
	const routes = [
		{ handle: "filesystem" },
		...introspectionData.routes,
		{
			src: "/(.*)",
			dest: "/"
		}
	];
	introspectionSpan.setAttributes({
		"introspection.success": "true",
		"introspection.routes": String(introspectionData.routes.length),
		"introspection.framework": introspectionData.frameworkSlug ?? ""
	});
	return {
		routes,
		framework,
		additionalFolders: introspectionData.additionalFolders ?? [],
		additionalDeps: introspectionData.additionalDeps ?? []
	};
};
const defaultResult = (args) => {
	return {
		routes: [{ handle: "filesystem" }, {
			src: "/(.*)",
			dest: "/"
		}],
		framework: getFramework(args)
	};
};
const getFramework = (args) => {
	try {
		let version$1;
		if (args.framework) {
			const frameworkLibPath = require.resolve(`${args.framework}`, { paths: [args.dir] });
			const findNearestPackageJson = (dir) => {
				const packageJsonPath = join(dir, "package.json");
				if (existsSync(packageJsonPath)) return packageJsonPath;
				const parentDir = dirname(dir);
				if (parentDir === dir) return;
				return findNearestPackageJson(parentDir);
			};
			const nearestPackageJsonPath = findNearestPackageJson(frameworkLibPath);
			if (nearestPackageJsonPath) version$1 = require(nearestPackageJsonPath).version;
		}
		return {
			slug: args.framework ?? "",
			version: version$1 ?? ""
		};
	} catch (error) {
		debug(`Error getting framework for ${args.framework}. Setting framework version to empty string.`, error);
		return {
			slug: args.framework ?? "",
			version: ""
		};
	}
};

//#endregion
//#region src/index.ts
const version = 2;
/** Non-empty Build Command from project settings / vercel.json (not the default `build` script). */
function hasExplicitBuildCommand(config) {
	const cmd = config.buildCommand ?? config.projectSettings?.buildCommand;
	return typeof cmd === "string" && cmd.trim().length > 0;
}
const build = async (args) => {
	if (typeof args.config?.handlerFunction === "string") throw new Error(`Named-function entrypoints are not supported for JavaScript services (got "${args.entrypoint}:${args.config.handlerFunction}"). Put each handler in its own file with a default export.`);
	const downloadResult = await downloadInstallAndBundle(args);
	const nodeVersion = await getNodeVersion(args.workPath, void 0, args.config, args.meta);
	const isBun = isBunVersion(nodeVersion);
	const builderName = "@vercel/backends";
	const span = args.span ?? new Span({ name: builderName });
	span.setAttributes({ "builder.name": builderName });
	const buildSpan = span.child("vc.builder.backends.build");
	return buildSpan.trace(async () => {
		const explicit = args.entrypoint && args.entrypoint !== "package.json" ? args.entrypoint : null;
		const entrypoint = explicit && existsSync(join(args.workPath, explicit)) ? explicit : await findEntrypointOrThrow(args.workPath);
		debug("Entrypoint", entrypoint);
		args.entrypoint = entrypoint;
		const cronEntries = getServiceCrons({
			service: args.service,
			entrypoint
		});
		const isCronService = cronEntries !== void 0;
		const userBuildResult = await maybeDoBuildCommand(args, downloadResult);
		const functionConfig = args.config.functions?.[entrypoint];
		if (functionConfig) {
			args.config.includeFiles = [...normalizeArray(args.config.includeFiles), ...normalizeArray(functionConfig.includeFiles)];
			args.config.excludeFiles = [...normalizeArray(args.config.excludeFiles), ...normalizeArray(functionConfig.excludeFiles)];
		}
		const rolldownResult = await rolldown({
			...args,
			span: buildSpan,
			defaultFormat: isCronService ? "cjs" : void 0
		});
		const introspectionPromise = !isCronService && rolldownResult.framework.slug === "hono" ? introspection({
			...args,
			span: buildSpan,
			files: rolldownResult.files,
			handler: rolldownResult.handler
		}) : Promise.resolve({
			routes: [],
			additionalFolders: [],
			additionalDeps: []
		});
		let typescriptPromise;
		if (hasExplicitBuildCommand(args.config)) {
			console.log(Colors.gray(`${Colors.bold(Colors.cyan("✓"))} Typecheck skipped ${Colors.gray("(Build Command is configured)")}`));
			typescriptPromise = Promise.resolve();
		} else typescriptPromise = typescript({
			entrypoint,
			workPath: args.workPath,
			span: buildSpan
		});
		const localBuildFiles = userBuildResult?.localBuildFiles.size > 0 ? userBuildResult?.localBuildFiles : rolldownResult.localBuildFiles;
		const files = userBuildResult?.files || rolldownResult.files;
		const handler = userBuildResult?.handler || rolldownResult.handler;
		const nftWorkPath = userBuildResult?.outputDir || args.workPath;
		await nft({
			...args,
			workPath: nftWorkPath,
			localBuildFiles,
			files,
			ignoreNodeModules: false,
			ignore: args.config.excludeFiles,
			conditions: isBun ? ["bun"] : void 0,
			span: buildSpan
		});
		try {
			await generateProjectManifest({
				workPath: args.workPath,
				nodeVersion,
				cliType: downloadResult.cliType,
				lockfilePath: downloadResult.lockfilePath,
				lockfileVersion: downloadResult.lockfileVersion,
				framework: rolldownResult.framework.slug || void 0,
				serviceType: args.service ? getReportedServiceType(args.service) : void 0
			});
		} catch (err) {
			debug(`Failed to write node manifest: ${err instanceof Error ? err.message : String(err)}`);
		}
		const baseDir = args.repoRootPath || args.workPath;
		const includeResults = await Promise.all(normalizeArray(args.config.includeFiles).map((pattern) => glob(pattern, baseDir)));
		for (const matched of includeResults) for (const [relPath, entry] of Object.entries(matched)) files[relPath] = entry;
		const introspectionResult = await introspectionPromise;
		await typescriptPromise;
		const functionConfigOverrides = await getLambdaOptionsFromFunction({
			sourceFile: entrypoint,
			config: args.config
		});
		const serviceRoutePrefix = normalizeServiceRoutePrefix(args.config?.routePrefix ?? args.service?.routePrefix);
		const shouldStripServiceRoutePrefix = !!serviceRoutePrefix && (typeof args.config?.serviceName === "string" || !!args.service);
		let lambdaFiles = files;
		let lambdaHandler = handler;
		if (isCronService && cronEntries) {
			const dispatched = await applyCronDispatch({
				files,
				handler,
				workPath: nftWorkPath,
				routes: buildCronRouteTable(cronEntries)
			});
			lambdaFiles = dispatched.files;
			lambdaHandler = dispatched.handler;
		} else if (shouldStripServiceRoutePrefix) {
			const shimmedLambda = await applyServiceVcInit({
				files,
				handler,
				workPath: nftWorkPath
			});
			lambdaFiles = shimmedLambda.files;
			lambdaHandler = shimmedLambda.handler;
		}
		const lambda = new NodejsLambda({
			runtime: nodeVersion.runtime,
			handler: lambdaHandler,
			files: lambdaFiles,
			framework: rolldownResult.framework,
			shouldAddHelpers: false,
			shouldAddSourcemapSupport: true,
			awsLambdaHandler: "",
			...functionConfigOverrides,
			shouldDisableAutomaticFetchInstrumentation: process.env.VERCEL_TRACING_DISABLE_AUTOMATIC_FETCH_INSTRUMENTATION === "1"
		});
		if (shouldStripServiceRoutePrefix && serviceRoutePrefix) lambda.environment = {
			...lambda.environment,
			VERCEL_SERVICE_ROUTE_PREFIX: serviceRoutePrefix,
			VERCEL_SERVICE_ROUTE_PREFIX_STRIP: "1"
		};
		const serviceName = typeof args.config?.serviceName === "string" && args.config.serviceName !== "" ? args.config.serviceName : void 0;
		const internalServiceFunctionPath = typeof serviceName === "string" && serviceName !== "" ? `/_svc/${serviceName}/index` : void 0;
		const internalServiceOutputPath = internalServiceFunctionPath?.slice(1);
		const remapRouteDestination = (route) => {
			const prefixedRoute = maybePrefixServiceRouteSource(route, serviceRoutePrefix);
			if (!internalServiceFunctionPath || !route.dest) return prefixedRoute;
			return {
				...prefixedRoute,
				dest: internalServiceFunctionPath
			};
		};
		const routes = isCronService ? [{ handle: "filesystem" }] : [
			{ handle: "filesystem" },
			...introspectionResult.routes.map(remapRouteDestination),
			{
				src: getServiceCatchallSource(serviceRoutePrefix),
				dest: internalServiceFunctionPath ?? "/"
			}
		];
		const output = internalServiceOutputPath ? { [internalServiceOutputPath]: lambda } : { index: lambda };
		for (const route of routes) if (route.dest) {
			if (route.dest === "/") continue;
			const outputPath = route.dest === internalServiceFunctionPath && internalServiceOutputPath ? internalServiceOutputPath : route.dest;
			output[outputPath] = lambda;
		}
		return {
			routes,
			output,
			...cronEntries ? { crons: cronEntries } : {}
		};
	});
};
const prepareCache = ({ repoRootPath, workPath }) => {
	return glob(defaultCachePathGlob, repoRootPath || workPath);
};
const normalizeArray = (value) => Array.isArray(value) ? value : value ? [value] : [];
const normalizeServiceRoutePrefix = (routePrefix) => {
	if (typeof routePrefix !== "string" || routePrefix === "" || routePrefix === ".") return;
	let normalized = routePrefix.startsWith("/") ? routePrefix : `/${routePrefix}`;
	if (normalized !== "/" && normalized.endsWith("/")) normalized = normalized.slice(0, -1);
	return normalized === "/" ? void 0 : normalized;
};
const maybePrefixServiceRouteSource = (route, routePrefix) => {
	if (!routePrefix || typeof route.dest !== "string" || !route.dest.startsWith("/")) return route;
	return {
		...route,
		src: getPrefixedRouteSource(route.src, route.dest, routePrefix)
	};
};
const getPrefixedRouteSource = (routeSource, routePath, routePrefix) => {
	if (!routeSource) return routeSource;
	if (routePath === routePrefix || routePath.startsWith(`${routePrefix}/`)) return routeSource;
	const escapedRoutePrefix = toRegexSource(routePrefix);
	if (routeSource.startsWith("^(?:")) return `^(?:${escapedRoutePrefix}${routeSource.slice(4)}`;
	if (routeSource.startsWith("^")) return `^${escapedRoutePrefix}${routeSource.slice(1)}`;
	return `${escapedRoutePrefix}${routeSource}`;
};
const getServiceCatchallSource = (routePrefix) => {
	if (!routePrefix) return "/(.*)";
	return `^${escapeForRegex(routePrefix)}(?:/(.*))?$`;
};
const escapeForRegex = (value) => value.replace(/[|\\{}()[\]^$+*?.]/g, "\\$&");
const toRegexSource = (value) => escapeForRegex(value).replaceAll("/", "\\/");

//#endregion
export { build, build$1 as cervelBuild, serve as cervelServe, diagnostics, findEntrypoint, findEntrypointOrThrow, getBuildSummary, introspectApp, nodeFileTrace, prepareCache, srvxOptions, version };
